File: android/app/src/main/java/com/openavc/panel/discovery/ServerValidator.kt
package com.openavc.panel.discovery

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.security.SecureRandom
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

/**
 * Performs a one-shot `/api/status` fetch against a candidate server and
 * returns a populated [ServerInfo] on success, or null if unreachable or
 * the response does not parse.
 *
 * HTTPS mode: fetches and pins the server's CA (auto-generate) or leaf
 * (provided cert) on first contact via [CertFetcher], then validates the
 * /api/status response under a [PinnedTrustManager]. After the server
 * returns its instance_id we re-pin under that key so the trust survives
 * IP changes across DHCP rotations.
 *
 * A plain-http probe of a TLS-enabled server is answered by the server's
 * HTTP-to-HTTPS redirect listener. That 3xx is treated as a TLS handoff:
 * validation retries over https against the same host on the redirected
 * port (see [tlsHandoffPort]), so scanning a printed http:// pair QR still
 * lands on a server that has since enabled HTTPS.
 */
object ServerValidator {

    private const val TAG = "ServerValidator"
    private const val CONNECT_TIMEOUT_MS = 3_000
    private const val READ_TIMEOUT_MS = 3_000

    internal fun buildUrl(host: String, port: Int, scheme: String, path: String): URL =
        URL("$scheme://$host:$port$path")

    /**
     * Decides whether a failed plain-http probe was answered by the server's
     * HTTP-to-HTTPS redirect listener. [HttpURLConnection] follows same-scheme
     * redirects on its own but refuses to cross http-to-https, so a 3xx that
     * surfaces here with an https Location means the server is alive on this
     * host and wants the conversation on its TLS port.
     *
     * Returns the https port to re-validate against, or null when the
     * response is not a TLS handoff. Only the Location's scheme and port are
     * honored; the caller keeps the host it just probed. The probed host is
     * the address that answered, and its bare-IP handshake serves the
     * self-signed chain that /api/certificate pinning can match — whereas a
     * server with a cloud-issued certificate puts its public DNS name in the
     * Location, which may not resolve without internet and presents a
     * CA-signed chain that a pinned certificate never matches.
     */
    internal fun tlsHandoffPort(code: Int, location: String?): Int? {
        if (code !in 300..399 || location.isNullOrBlank()) return null
        val target = try {
            URL(location.trim())
        } catch (e: MalformedURLException) {
            // Relative or unparseable Location — not a cross-scheme handoff.
            return null
        }
        // URL() is lenient ("https://" parses with an empty host) — a
        // hostless Location is garbage, not a handoff.
        if (!target.protocol.equals("https", ignoreCase = true) || target.host.isNullOrEmpty()) {
            return null
        }
        return if (target.port != -1) target.port else 443
    }

    suspend fun validate(
        context: Context,
        host: String,
        port: Int,
        scheme: String = "http",
    ): ServerInfo? = withContext(Dispatchers.IO) {
        val statusUrl = buildUrl(host, port, scheme, "/api/status")
        val pinnedCert = if (scheme == "https") {
            CertFetcher(context).fetchAndPin(host, port, instanceId = null)
                ?: run {
                    Log.d(TAG, "could not fetch/pin cert for https://$host:$port")
                    return@withContext null
                }
        } else null

        val conn = openConnection(statusUrl, pinnedCert) ?: return@withContext null
        try {
            conn.connectTimeout = CONNECT_TIMEOUT_MS
            conn.readTimeout = READ_TIMEOUT_MS
            conn.requestMethod = "GET"
            conn.setRequestProperty("Accept", "application/json")
            val code = conn.responseCode
            if (code !in 200..299) {
                val handoffPort = if (scheme == "http") {
                    tlsHandoffPort(code, conn.getHeaderField("Location"))
                } else null
                if (handoffPort != null) {
                    Log.d(TAG, "TLS handoff from $statusUrl to https port $handoffPort")
                    return@withContext validate(context, host, handoffPort, "https")
                }
                Log.d(TAG, "status $code from $statusUrl")
                return@withContext null
            }
            val body = conn.inputStream.bufferedReader().use { it.readText() }
            val json = JSONObject(body)
            val name = json.optString("project_name")
                .ifEmpty { json.optString("name") }
                .ifEmpty { host }
            val version = json.optString("version")
            val instanceId = json.optString("instance_id")
            // Re-pin under instance_id so the pin survives DHCP-driven IP changes.
            if (pinnedCert != null && instanceId.isNotBlank()) {
                CertTrustStore(context).pin(
                    instanceId = instanceId,
                    hostPort = CertTrustStore.hostPortKey(host, port),
                    pemBytes = CertFetcher.pemEncode(pinnedCert).toByteArray(Charsets.UTF_8),
                )
            }
            ServerInfo(
                name = name,
                instanceId = instanceId,
                host = host,
                port = port,
                version = version,
                panelUrl = "$scheme://$host:$port/panel",
                scheme = scheme,
            )
        } catch (e: Exception) {
            Log.d(TAG, "validate failed for $host:$port: ${e.message}")
            null
        } finally {
            conn.disconnect()
        }
    }

    suspend fun ping(
        context: Context,
        host: String,
        port: Int,
        scheme: String = "http",
        instanceId: String? = null,
    ): Boolean = withContext(Dispatchers.IO) {
        val url = buildUrl(host, port, scheme, "/api/health")
        val pinnedCert = if (scheme == "https") {
            CertTrustStore(context).lookup(instanceId, CertTrustStore.hostPortKey(host, port))
                ?: CertFetcher(context).fetchAndPin(host, port, instanceId)
        } else null
        val conn = openConnection(url, pinnedCert) ?: return@withContext false
        try {
            conn.connectTimeout = CONNECT_TIMEOUT_MS
            conn.readTimeout = READ_TIMEOUT_MS
            conn.requestMethod = "GET"
            conn.responseCode in 200..299
        } catch (e: Exception) {
            false
        } finally {
            conn.disconnect()
        }
    }

    private fun openConnection(
        url: URL,
        pinnedCert: java.security.cert.X509Certificate?,
    ): HttpURLConnection? {
        val conn = url.openConnection() as? HttpURLConnection ?: return null
        if (pinnedCert != null && conn is HttpsURLConnection) {
            val ctx = SSLContext.getInstance("TLS")
            ctx.init(
                null,
                arrayOf<X509TrustManager>(PinnedTrustManager(listOf(pinnedCert))),
                SecureRandom(),
            )
            conn.sslSocketFactory = ctx.socketFactory
            // The server cert's SANs cover localhost + the LAN IPs we advertise,
            // but a user can still hit it under a hostname we didn't enumerate.
            // Pinning the cert is the actual trust check; hostname is moot.
            conn.hostnameVerifier = HostnameVerifier { _, _ -> true }
        }
        return conn
    }
}

File: ios/OpenAVCPanel/Discovery/ServerValidator.swift
import Foundation
import Security

/// One-shot `/api/status` probe against a candidate server.
///
/// Returns a populated `ServerInfo`, or nil if the server is unreachable or the
/// response does not parse. Mirrors Android's `ServerValidator.kt`.
///
/// HTTPS mode: fetches and pins the server's CA (auto-generate) or leaf
/// (provided) on first contact through `CertFetcher`, then reads `/api/status`
/// under that pin. Once the server returns its `instance_id` we re-pin under
/// that key, so trust survives a DHCP address change.
enum ServerValidator {

    private static let timeout: TimeInterval = 3

    static func url(host: String, port: Int, scheme: String, path: String) -> URL? {
        URL(string: "\(scheme)://\(host):\(port)\(path)")
    }

    /// Decides whether a failed plain-http probe was answered by the server's
    /// HTTP-to-HTTPS redirect listener.
    ///
    /// Returns the https port to re-validate against, or nil when this is not a
    /// TLS handoff. **Only the Location's scheme and port are honoured — the
    /// caller keeps the host it just probed.** That is deliberate: the probed
    /// address is what answered, and its bare-IP handshake serves the
    /// self-signed chain that `/api/certificate` pinning can match. A server
    /// with a cloud-issued certificate puts its public DNS name in the Location
    /// instead, which may not resolve without internet and presents a CA-signed
    /// chain no pinned certificate will ever match.
    static func tlsHandoffPort(code: Int, location: String?) -> Int? {
        guard (300..<400).contains(code) else { return nil }
        guard let location, !location.trimmingCharacters(in: .whitespaces).isEmpty else { return nil }
        guard let target = URL(string: location.trimmingCharacters(in: .whitespaces)) else { return nil }
        guard target.scheme?.lowercased() == "https" else { return nil }
        guard let host = target.host, !host.isEmpty else { return nil }
        return target.port ?? 443
    }

    static func validate(
        host: String,
        port: Int,
        scheme: String = "http",
        trustStore: CertTrustStore = CertTrustStore()
    ) async -> ServerInfo? {
        guard let statusUrl = url(host: host, port: port, scheme: scheme, path: "/api/status") else {
            return nil
        }

        var pinned: SecCertificate?
        if scheme == "https" {
            pinned = await CertFetcher(trustStore: trustStore)
                .fetchAndPin(host: host, port: port, instanceId: nil)
            guard pinned != nil else { return nil }
        }

        let delegate = PinnedSessionDelegate(pinned: pinned)
        let session = makeSession(delegate: delegate)
        defer { session.finishTasksAndInvalidate() }

        var request = URLRequest(url: statusUrl)
        request.httpMethod = "GET"
        request.timeoutInterval = timeout
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let data: Data
        let http: HTTPURLResponse
        do {
            let (body, response) = try await session.data(for: request)
            guard let asHttp = response as? HTTPURLResponse else { return nil }
            data = body
            http = asHttp
        } catch {
            return nil
        }

        guard (200..<300).contains(http.statusCode) else {
            // The delegate refuses redirects, so a 3xx surfaces here intact.
            if scheme == "http",
               let handoff = tlsHandoffPort(
                   code: http.statusCode,
                   location: http.value(forHTTPHeaderField: "Location")
               ) {
                return await validate(
                    host: host, port: handoff, scheme: "https", trustStore: trustStore
                )
            }
            return nil
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }

        let name = (json["project_name"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            ?? (json["name"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            ?? host
        let version = json["version"] as? String ?? ""
        let instanceId = json["instance_id"] as? String ?? ""

        // Re-pin under instance_id so the pin survives DHCP-driven IP changes.
        if let pinned, !instanceId.isEmpty {
            trustStore.pin(
                instanceId: instanceId,
                hostPort: CertTrustStore.hostPortKey(host: host, port: port),
                pem: CertTrustStore.pemEncode(pinned)
            )
        }

        return ServerInfo(
            name: name,
            instanceId: instanceId,
            host: host,
            port: port,
            version: version,
            panelUrl: "\(scheme)://\(host):\(port)/panel",
            scheme: scheme
        )
    }

    /// Cheap liveness check for a server we already know, used on launch to
    /// decide whether discovery can be skipped.
    static func ping(
        host: String,
        port: Int,
        scheme: String = "http",
        instanceId: String? = nil,
        trustStore: CertTrustStore = CertTrustStore()
    ) async -> Bool {
        guard let healthUrl = url(host: host, port: port, scheme: scheme, path: "/api/health") else {
            return false
        }

        var pinned: SecCertificate?
        if scheme == "https" {
            let hostPort = CertTrustStore.hostPortKey(host: host, port: port)
            if let stored = trustStore.lookup(instanceId: instanceId, hostPort: hostPort) {
                pinned = stored
            } else {
                pinned = await CertFetcher(trustStore: trustStore)
                    .fetchAndPin(host: host, port: port, instanceId: instanceId)
            }
            guard pinned != nil else { return false }
        }

        let session = makeSession(delegate: PinnedSessionDelegate(pinned: pinned))
        defer { session.finishTasksAndInvalidate() }

        var request = URLRequest(url: healthUrl)
        request.httpMethod = "GET"
        request.timeoutInterval = timeout

        guard let (_, response) = try? await session.data(for: request),
              let http = response as? HTTPURLResponse
        else { return false }
        return (200..<300).contains(http.statusCode)
    }

    private static func makeSession(delegate: URLSessionDelegate) -> URLSession {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout
        return URLSession(configuration: config, delegate: delegate, delegateQueue: nil)
    }
}
